[Management] Address UI searching bug with index pattern creation

chrisronline opened the pull request:
## Overview
On the index pattern creation wizard, if the user types in a character into the input field, we have code that will automatically append a `*` character to the end of what they typed and then move the cursor back to right after the typed letter. This experience came from [discussions on the original PR](https://github.com/elastic/kibana/pull/13454#issuecomment-321850312).

Unfortunately, there is a bug with how we're transferring state between the angular directives where we weren't properly updating the state before the parent directive accessed it, leading to an initial query against the typed character, instead of the typed character plus the `*`. 

For super users, this isn't a big deal because they aren't any security gates against searching for indices, but it becomes a problem for restricted users who aren't able to query for certain indices.

Take this sample of queries for a restricted user who can only access the logstash index:

```
POST l/_search -> 403
POST *l/_search -> 200
POST l*/_search -> 200
POST *l*/_search -> 200
```

The UI bug caused a search for just `l` instead of `l*` which resulted in 403 errors in the browser.

This PR addresses that and separates out the code into a separate directive. It also functions off `keydown` instead of an angular watch due to issues mutating that value post watch firing. 

It's very possible this is a hacky and there is a better way, so please let me know!

## TODO
- Tests are hard for this kind of experience. Any recommendations?

bmcconaghy commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/step_index_pattern/append_wildcard.js:
```diff
@@ -0,0 +1,53 @@
+        }
+      }
+
+      $elem.on('keydown', (e) => {
```
Maybe a little concerned that this will not always work as intended (all work completing before watcher has a chance to fire).  Maybe that is unfounded though.

chrisronline replied:
It's a good point and something I just assumed. I figured that the digest cycle doesn't kickoff until all interactions are "done", and I assumed that any listeners directly listening to UI interacts will be handled prior to anything angular does, but I could be mistaken there

bmcconaghy replied:
Methods run to completion in JS, so this should be good, just being paranoid.

bmcconaghy commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/step_index_pattern/append_wildcard.js:
```diff
@@ -0,0 +1,53 @@
+        event.preventDefault();
+        $elem.val(val);
+        if (moveSelectionRange) {
+          $elem[0].setSelectionRange(1, 1);
```
I had to think for a minute to grok why this line is doing what it does.  Maybe parameter could be named better to make that clear.

chrisronline replied:
On it!

bmcconaghy commented:
LGTM

kobelb commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/step_index_pattern/append_wildcard.js:
```diff
@@ -0,0 +1,53 @@
+      }
+
+      $elem.on('keydown', (e) => {
+        const charStr = String.fromCharCode(e.keyCode);
```
MDN warns against using [KeyboardEvent.keyCode](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) and recommends using [KeyboardEvent.key](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key). However, Chrome 51 was the first to introduce `KeyboardEvent.key`, so it'd be nice if this "gracefully degrades" in unsupported versions, perhaps by simply not intelligently adding the `*`.

kobelb replied:
This might be related to the following behavior, if I type `Command + R` (normally refreshes the page in Chrome on macOS) it's inserting a lower-case "r":

![command-r-hijacking](https://user-images.githubusercontent.com/627123/32173212-97e75ac6-bd54-11e7-9edd-52d8829763b4.gif)

chrisronline replied:
Nice call! Addressing the first one fixes the second one

kobelb commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/step_index_pattern/append_wildcard.js:
```diff
@@ -0,0 +1,53 @@
+
+        if (indexPatternName && indexPatternName.length === 1) {
+          if (indexPatternName === '*') {
+            if ($scope.appendedWildcard) {
```
This causes an interesting behavior, when the page first loads you're allowed to put `*` as the index-pattern, but if I create an index pattern where we append the `*` and `$scope.appendedWildcard === true`, if I try to type `*` again, it won't do anything, but if I press it again, it'll let me. This appears to be behavior that was carried over from the previous code, so if you'd prefer to create a separate PR/Issue to address this, I won't let it hold up my approval.

![asterix](https://user-images.githubusercontent.com/627123/32173105-4276bd52-bd54-11e7-881d-08f05b6d368e.gif)

chrisronline commented:
@kobelb Thanks for the great thoughts! I fixed a couple things and refactored some of the code to hopefully avoid some of the confusion/issues. Lemme know

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/step_index_pattern/append_wildcard.js:
```diff
@@ -0,0 +1,50 @@
+          return;
+        }
+
+        // If the user is holdinng down ctrl/cmd, they are performing some shortcut
```
Bah, typo in the comment

chrisronline commented on src/core_plugins/kibana/public/management/sections/indices/create_index_pattern_wizard/step_index_pattern/append_wildcard.js:
```diff
@@ -0,0 +1,50 @@
+
+        let indexPatternName = $elem.val() + e.key;
+        if (indexPatternName && indexPatternName.length === 1) {
+          if (indexPatternName !== '*') {
```
This can be combined in the above if

chrisronline commented:
Backport

6.x: https://github.com/elastic/kibana/commit/076cea169475823b5e8ba5330dc6a6eee8d9a366
